File: Antidote/CopyLabel.m
//
//  CopyLabel.m
//  Antidote
//
//  Created by Dmitry Vorobyov on 19.07.14.
//  Copyright (c) 2014 dvor. All rights reserved.
//

#import "CopyLabel.h"

@implementation CopyLabel

#pragma mark -  Lifecycle

- (id)initWithFrame:(CGRect)frame
{
    self = [super initWithFrame:frame];
    if (self) {
        self.userInteractionEnabled = YES;

        UITapGestureRecognizer *tapGR = [[UITapGestureRecognizer alloc] initWithTarget:self
                                                                                action:@selector(tapGesture:)];

        [self addGestureRecognizer:tapGR];
    }
    return self;
}

#pragma mark -  Gestures

- (void)tapGesture:(UITapGestureRecognizer *)tapGR
{
    [self becomeFirstResponder];

    UIMenuController *menu = [UIMenuController sharedMenuController];
    [menu setTargetRect:self.frame inView:self.superview];
    [menu setMenuVisible:YES animated:YES];
}

#pragma mark -  Private

- (void)copy:(id)sender
{
    UIPasteboard *pasteboard = [UIPasteboard generalPasteboard];
    pasteboard.string = self.text;
}

- (BOOL)canPerformAction:(SEL)action withSender:(id)sender
{
    return (action == @selector(copy:));
}

- (BOOL)canBecomeFirstResponder
{
    return YES;
}

@end

File: Antidote/CopyLabel.swift
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at http://mozilla.org/MPL/2.0/.

import UIKit

class CopyLabel: UILabel {
    var copyable = true {
        didSet {
            recognizer.isEnabled = copyable
        }
    }

    fileprivate var recognizer: UITapGestureRecognizer!

    override init(frame: CGRect) {
        super.init(frame: frame)

        isUserInteractionEnabled = true

        recognizer = UITapGestureRecognizer(target: self, action: #selector(CopyLabel.tapGesture))
        addGestureRecognizer(recognizer)
    }

    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
}

// MARK: Actions
extension CopyLabel {
    @objc func tapGesture() {
        // This fixes issue with calling UIMenuController after UIActionSheet was presented.
        let appDelegate = UIApplication.shared.delegate as! AppDelegate
        appDelegate.window?.makeKey()

        becomeFirstResponder()

        let menu = UIMenuController.shared
        menu.setTargetRect(frame, in: superview!)
        menu.setMenuVisible(true, animated: true)
    }
}

// MARK: Copying
extension CopyLabel {
    override func copy(_ sender: Any?) {
        UIPasteboard.general.string = text
    }

    override func canPerformAction(_ action: Selector, withSender sender: Any?) -> Bool {
        return action == #selector(copy(_:))
    }

    override var canBecomeFirstResponder : Bool {
        return true
    }
}
